### PR TITLE
Added support for function annotations

### PR DESCRIPTION
**Pull request by hearot**
You are now able to experience a very new way to create your own commands. Instead of doing multiple checks & conversions on parameters, you may now use annotations (see [PEP 3107](https://www.python.org/dev/peps/pep-3107/)) in order to convert automatically them. 

- Example: https://gist.github.com/hearot/e00bd75312a781c4a20db1131585cd38

### CODE REVIEW

**Comment by hearot** [comment]
I haven't updated the docs yet. I'll update them as soon as possible 💯

**Comment by matteob99** [comment]
missing doc and test

File: docs/api/bot.rst [hunk 1]
```diff
@@ -289,18 +289,23 @@ components.
...
       to the ``/help`` message. The decorated function will be called with
-      three parameters:
+      these parameters:
 
       * A ``chat`` parameter with the representation of the chat in which the
         message was sent (an instance of :py:class:`botogram.Chat`)
       * A ``message`` parameter with the representation of the received
         message (an instance of :py:class:`botogram.Message`)
       * An ``args`` parameter with the list of parsed arguments
+      * Any other arguments which are required by the bot to call the command.
```
Comment (matteob99) [inline docs/api/bot.rst:299]: missing versionadded

File: docs/api/bot.rst [hunk 2]
```diff
@@ -289,18 +289,23 @@ components.
...
+        Those special arguments are automatically converted if a *type annotation*
+        is provided. Default parameters are supported as well
 
       If you put a docstring on the decorated function, that will be used as
-      extended description of the command in the ``/help`` command. Also, if
-      you don't want this command to appear in the ``/help``, you can set the
-      ``hidden`` argument to ``True``.
+      extended description of the command in the ``/help`` command with the
+      required arguments as well.
```
Comment (matteob99) [inline docs/api/bot.rst:305]: missing versionadded

File: docs/api/components.rst [hunk 3]
```diff
@@ -248,19 +248,23 @@ about how to create them in the ":ref:`custom-components`" chapter.
...
       * A ``message`` parameter with the representation of the received
         message (an instance of :py:class:`botogram.Message`)
       * An ``args`` parameter with the list of parsed arguments
+      * Any other arguments which are required by the bot to call the command.
+        Those special arguments are automatically converted if a *type annotation*
+        is provided. Default parameters are supported as well
 
       If you put a docstring on the provided function, that will be used as
-      extended description of the command in the ``/help`` command.
+      extended description of the command in the ``/help`` command with the
```
Comment (matteob99) [inline docs/api/components.rst:263]: missing versionadded

File: docs/bot-structure.rst [hunk 4]
```diff
@@ -70,6 +70,40 @@ If you need to store global state check out :ref:`shared memory
...
+   def say_hello_command(chat, from_who: str, to_who: str):
+      chat.send("%s says hello to %s!" % (from_who, to_who))
+
+ If you send ``/say_hello foo bot``, you will receive *foo says hello to bot* as a message.
+
+As you can see, you can use *type annotations*! If you use them, the bot will automatically
+convert the arguments to the provided types.
+
+You can use *default arguments* as well.
+
```
Comment (matteob99) [inline docs/bot-structure.rst:106]: missing versionadded

File: requirements-build.in [hunk 5]
```diff
@@ -1,2 +1,3 @@
-Babel
+babel
+typing
```
Comment (matteob99) [inline requirements-build.in:2]: why you change this file?

File: requirements-build.txt [hunk 6]
```diff
@@ -1,9 +1,10 @@
 #
 # This file is autogenerated by pip-compile
-# Make changes in requirements-build.in, then run this to update:
+# To update, run:
```
Comment (matteob99) [inline requirements-build.txt:3]: why you change this file?

File: requirements-docs.txt [hunk 7]
```diff
@@ -1,13 +1,14 @@
 #
 # This file is autogenerated by pip-compile
-# Make changes in requirements-docs.in, then run this to update:
+# To update, run:
```
Comment (matteob99) [inline requirements-docs.txt:3]: why you change this file?

File: requirements-lint.txt [hunk 8]
```diff
@@ -1,6 +1,6 @@
 #
 # This file is autogenerated by pip-compile
-# Make changes in requirements-lint.in, then run this to update:
+# To update, run:
```
Comment (matteob99) [inline requirements-lint.txt:3]: why you change this file?

File: requirements-test.in [hunk 9]
```diff
@@ -1,2 +1,3 @@
 pytest
 responses
+typing
```
Comment (matteob99) [inline requirements-test.in:3]: why you change this file?

File: requirements-test.txt [hunk 10]
```diff
@@ -1,12 +1,18 @@
 #
 # This file is autogenerated by pip-compile
-# Make changes in requirements-test.in, then run this to update:
+# To update, run:
```
Comment (matteob99) [inline requirements-test.txt:3]: why you change this file?

File: requirements-tools.txt [hunk 11]
```diff
@@ -1,6 +1,6 @@
 #
 # This file is autogenerated by pip-compile
-# Make changes in requirements-tools.in, then run this to update:
+# To update, run:
```
Comment (matteob99) [inline requirements-tools.txt:3]: why you change this file?

File: botogram/defaults.py [hunk 12]
```diff
@@ -95,8 +95,13 @@ def _help_generic_message(self, bot, commands):
                 summary = escape_html(command.summary)
                 if summary is None:
                     summary = "<i>%s</i>" % bot._("No description available.")
-                message.append("/%s <code>-</code> %s" %
-                               (command.name, summary))
+                parameters_list = command.parameters_list
```
Comment (matteob99) [inline botogram/defaults.py:98]: why you create new variable instead of using the same

File: .gitignore [hunk 13]
```diff
@@ -9,3 +9,7 @@ __pycache__
 /.cache
 
 /.netlify/build
+
+\.idea/
```
Comment (matteob99) [inline .gitignore:13]: don't add this folder to .gitignore if you need this add to .git/info/exclude

File: docs/api/bot.rst (hunk 1 above)
Comment (hearot) [reply docs/api/bot.rst:299]: Added in 4053df7

File: docs/api/bot.rst (hunk 2 above)
Comment (hearot) [reply docs/api/bot.rst:305]: Added in 4f32464

File: docs/api/components.rst (hunk 3 above)
Comment (hearot) [reply docs/api/components.rst:263]: Added in 7ba3e2f

File: docs/bot-structure.rst (hunk 4 above)
Comment (hearot) [reply docs/bot-structure.rst:106]: Added in 21a82db

File: requirements-build.in (hunk 5 above)
Comment (hearot) [reply requirements-build.in]: 995f068 made it for me

File: requirements-build.txt (hunk 6 above)
Comment (hearot) [reply requirements-build.txt]: 995f068 made it for me

File: requirements-docs.txt (hunk 7 above)
Comment (hearot) [reply requirements-docs.txt]: 995f068 made it for me

File: requirements-lint.txt (hunk 8 above)
Comment (hearot) [reply requirements-lint.txt]: 995f068 made it for me

File: requirements-test.in (hunk 9 above)
Comment (hearot) [reply requirements-test.in:3]: 995f068 made it for me

File: requirements-test.txt (hunk 10 above)
Comment (hearot) [reply requirements-test.txt]: 995f068 made it for me

File: requirements-tools.txt (hunk 11 above)
Comment (hearot) [reply requirements-tools.txt]: 995f068 made it for me

File: botogram/defaults.py (hunk 12 above)
Comment (hearot) [reply botogram/defaults.py]: Done in c4f983b. Good ol' days

File: .gitignore (hunk 13 above)
Comment (hearot) [reply .gitignore]: Done in e90bc94

**Comment by hearot** [comment]
👀 @MarcoBuster

**Comment by hearot** [comment]
@hearot L'abbiamo perso, ormai

**Comment by MarcoBuster** [comment]
@hearot @hearot I'm more alive than ever

**Comment by giuseppeM99** [comment]
are you gonna live forever?

Il sab 18 lug 2020, 22:11 Marco Aceti <notifications@github.com> ha scritto:

**Comment by hearot** [comment]
@giuseppeM99 👀 https://www.youtube.com/watch?v=dQw4w9WgXcQ

**Comment by giuseppeM99** [comment]
@giuseppeM99 .del

**Comment by MarcoBuster** [comment]
Sorry for the long waiting time! Thank you for your PR :+1:

**Comment by hearot** [comment]
@MarcoBuster OMG

**Pull request merged**